src/client/chat.js

```diff
@@ -0,0 +1,46 @@
+const states = require('../states')
+
+module.exports = (client, options) => {
+  const mcData = require('minecraft-data')(client.version)
+  const players = {} // 1.19+
+
+  function onChat (packet, isMessageFromServer = false) {
+    let message = packet.message ?? packet.unsignedChatContent ?? packet.signedChatContent
+    if (isMessageFromServer) message = packet.content
+
+    let verified = isMessageFromServer ? null : false // default to false if this is a system chat
+    if (!isMessageFromServer && client.profileKeys != null && packet.signature) {
+      verified = client.verifyMessage(players[packet.senderUuid].publicKey, packet)
+    }
+    client.emit('chat_received', { verified, message, isMessageFromServer })
```

Comment:
Why not a chatMessage and serverMessage event ? In mineflayer we don't handle both in one function and the event emitted has a type in it. If chat abstraction is moving to nmp just copy over the existing mineflayer implementation / events